cli/cmd/awsiam.go

```diff
@@ -0,0 +1,325 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+	"github.com/aws/aws-sdk-go-v2/service/iam/types"
+	"github.com/lacework/go-sdk/lwrunner"
+)
+
+// setupSSMAccess sets up an IAM role for SSM and attaches it to
+// the machine's instance profile. Takes role name as argument;
+// pass the empty string to create a new role.
+// Then creates SSM document.
+func setupSSMAccess(cfg aws.Config, roleName string, token string) (types.Role, types.InstanceProfile, error) {
+	cli.Log.Debugw("setting up role", "passed roleName", roleName)
+	role, err := setupSSMRole(cfg, roleName)
+	if err != nil {
+		return role, types.InstanceProfile{}, err
+	}
+
+	err = attachSSMPoliciesToRole(cfg, role)
+	if err != nil {
+		return role, types.InstanceProfile{}, err
+	}
+
+	// Create instance profile and add the role to it
+	instanceProfile, err := setupInstanceProfile(cfg, role)
+	if err != nil {
+		return role, instanceProfile, err
+	}
+
+	return role, instanceProfile, nil
+}
+
+// teardownSSMAccess destroys all the infra created during the execution of this program.
+// Specifically, this function:
+// - Removes the role from the instance profile
+// - Deletes the instance profile
+// - Detaches all managed policies from the role (assumes no inline policies attached)
+// - Deletes the role
+func teardownSSMAccess(cfg aws.Config, role types.Role, instanceProfile types.InstanceProfile, byoRoleName string) error {
+	c := iam.New(iam.Options{
+		Credentials: cfg.Credentials,
+		Region:      cfg.Region,
+	})
+
+	cli.Log.Debug("only destroying instance profile if it's ours")
+	if *instanceProfile.InstanceProfileName == instanceProfileName {
+		cli.Log.Debugw("removing role from instance profile", "role", role, "instance profile", instanceProfile)
+		removeInput := &iam.RemoveRoleFromInstanceProfileInput{
+			InstanceProfileName: instanceProfile.InstanceProfileName,
+			RoleName:            role.RoleName,
+		}
+		_, err := c.RemoveRoleFromInstanceProfile(context.Background(), removeInput)
+		if err != nil {
+			return err
+		}
+
+		cli.Log.Debugw("deleting instance profile", "instance profile", instanceProfile)
+		deleteProfileInput := &iam.DeleteInstanceProfileInput{
+			InstanceProfileName: instanceProfile.InstanceProfileName,
+		}
+		_, err = c.DeleteInstanceProfile(context.Background(), deleteProfileInput)
+		if err != nil {
+			return err
+		}
+	}
+
+	if byoRoleName != "" || *role.RoleName != roleName {
+		return nil // we didn't create this role, we should not delete it
+	}
```

Comment:
Same: examining an explicit tag rather than an implicit property of the script seems safer.

Reply:
Done with 162cf4a1ca3a9722a25341408478d7f6edaf476c